README.md
[![Build Status](https://travis-ci.org/globocom/thumbor-client-ex.svg?branch=master)](https://travis-ci.org/globocom/thumbor-client-ex) [![HitCount](http://hits.dwyl.io/globocom/thumbor-client-ex.svg)](http://hits.dwyl.io/globocom/thumbor-client-ex)

# ThumborClient

This package is a client to generate URLs to [Thumbor](https://github.com/thumbor/thumbor) using Elixir language.

## Installation

The package can be installed by adding `thumbor_client` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:thumbor_client, "~> 0.4.0"}
  ]
end
```

## Usage

### In safe mode

```elixir
iex> client = ThumborClient.client("my_secret_token")
iex> client.(%{image: "path/to/image.jpg", width: 500, height: 500})
"1_6x25QaeExcQVmtuNyrr_lOs-0=/500/500/path/to/image.jpg"
```

### In unsafe mode

```elixir
iex> client = ThumborClient.client()
iex> client.("%{image: "path/to/image.jpg", width: 500, height: 500})
"unsafe/500/500/path/to/image.jpg"
```

### Another way to generate

The method `client("key")` is recommended if you will generate multiple images in same function.
If you prefer, you can call the method generate without this HOF.

```elixir
iex> ThumborClient.generate(%{image: "image.jpg", width: 500, height: 500}, "my_secret_token")
# The last parameter is optional, if not exist should use unsafe mode
```

## Options

|      Option       | Default Value |                                                                                               Description                                                                                               |
|-------------------|---------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| trim: (bool)      | false         | Removes surrounding space in image using top-left pixel color unless specified otherwise                                                                                                               |
| meta: (bool)      | false         | Instead of get the image, get all meta data informations in image returning using json                                                                                                                  |
| fit: (atom\|nil)   | nil           | The fit argument specifies that the image should not be auto-cropped and auto-resized to be EXACTLY the specified size. Possible values: :fit_in, :adaptive_fit_in, :full_fit_in, :adaptive_full_fit_in |
| width: (integer)  | 0             | Final width of image                                                                                                                                                                                    |
| height: (integer) | 0             | Final height of image                                                                                                                                                                                   |
| flip: (bool)      | false         | Flip image horizontaly                                                                                                                                                                                  |
| flop: (bool)      | false         | Flip image verticaly                                                                                                                                                                                    |
| halign: (atom)    | :center       | Orientation to crop horizontaly. Possible values: `:left`, `:center`, `:right`                                                                                                                                |
| valign: (atom)    | :center       | Orientation to crop verticaly. Possible values: `:top`, `:center`, `:bottom`                                                                                                                                  |
| smart: (bool)     | false         | Use Thumbor algorithms to crop using facial recognition process                                                                                                                                         |
| filters: (list)   | []            | Adding filters to image. More details bellow                                                                                                                                                            |
| image: (string)   | nil           | Path of image. Can be external.                                                                                                                                                                         |

## Filters

You can see a big list of filters in official [Thumbor documentation](https://github.com/thumbor/thumbor/wiki/Filters). You must set a list of strings with values.

Examples of usage:

#### Brightness

```elixir
ThumborClient.generate(%{filters: ["brightness(40)"], width: 400, height: 300, path: "/path/image.jpg"})
```
#### Blur

```elixir
ThumborClient.generate(%{filters: ["blur(40)"], width: 400, height: 300, path: "/path/image.jpg"})
```

#### Grayscale

```elixir
ThumborClient.generate(%{filters: ["grayscale()"], width: 400, height: 300, path: "/path/image.jpg"})
```

#### Multiple filters

```elixir
ThumborClient.generate(%{filters: ["grayscale()", "rotate(90)", "saturate(20)"], width: 400, height: 300, path: "/path/image.jpg"})
```

## Docs

The docs can be found at [https://hexdocs.pm/thumbor_client](https://hexdocs.pm/thumbor_client).



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.4.0] - 2018-05-05
### Added
- Add fit param
- Refactors
- Add documentation of options in README file

## [0.3.0] - 2018-05-03
### Added
- Tests
- Support to filters
- Add support to flip image

## [0.2.0] - 2018-05-02
### Added
- Support to smart images
- New defmodule to manipulate the url
- Method generate instead of HOF to create client
- Add support to build url with trim
- Add support to build url with horizontal align (halign)
- Add support to build url with vertial align (valign)
- Add support to build url with smart parameter
- Add support to get meta data from image
- More tests
- Add doctest to ThumborClient.Url

## [0.1.0] - 2018-05-02
### Added
- Created the module ThumborClient
- Generated documentation and publish in Hex
- Update Readme


test/thumbor_client_test.exs
defmodule ThumborClientTest do
  use ExUnit.Case

  doctest ThumborClient

  @key "123"
  @path "path/to/image.jpg"

  test "#encrypt_to_thumbor when pass a key" do
    assert ThumborClient.encrypt_to_thumbor(@key, @path) == "CXYhsuabkwiYi7STg9P1EHMnVyY="
  end

  test "#encrypt_to_thumbor when using unsafe mode" do
    assert ThumborClient.encrypt_to_thumbor(false, @path) == "unsafe"
  end

  test "#generate with same value of client" do
    client = ThumborClient.client(@key)
    options = %{width: 20, height: 10, image: @path, smart: true}
    assert client.(options) === ThumborClient.generate(options, @key)
  end

  test "#generate with key and trim" do
    options = %{width: 20, height: 10, image: @path, trim: true}
    assert ThumborClient.generate(options, @key) == "VKudNRI0uT6Ksa_QaVmW3ZAHILs=/trim/20x10/path/to/image.jpg"
  end

  test "#generate with horizontal align" do
    options = %{width: 20, height: 10, image: @path, halign: :left}
    assert ThumborClient.generate(options, @key) == "nEL0THUQpPIvQMo-duIr6gakHcw=/20x10/left/path/to/image.jpg"
  end

  test "#generate with vertical align" do
    options = %{width: 20, height: 10, image: @path, valign: :bottom}
    assert ThumborClient.generate(options, @key) == "cpfZnhhYgp0qmxgfVFCoNoZ1jkU=/20x10/bottom/path/to/image.jpg"
  end

  test "#generate with horizontal and vertical align" do
    options = %{width: 20, height: 10, image: @path, halign: :right, valign: :bottom}
    assert ThumborClient.generate(options, @key) == "u2mOj-6R2bjtCs8YzGxrJlbt37c=/20x10/right/bottom/path/to/image.jpg"
  end

  test "#generate with meta" do
    options = %{width: 20, height: 10, image: @path, meta: true}
    assert ThumborClient.generate(options, @key) == "Mijch4AjWk86uc5tzfHS1ne9pXA=/meta/20x10/path/to/image.jpg"
  end

  test "#generate with filter" do
    options = %{width: 20, height: 10, image: @path, filters: ["rotate(90)"]}
    assert ThumborClient.generate(options, @key) == "369I9wW_9-SfrWUmHgnrVoxbyiA=/20x10/filters:rotate(90)/path/to/image.jpg"
  end
end


test/thumbor_client_url_test.exs
defmodule ThumborClientUrlBuilderTest do
  use ExUnit.Case
  doctest ThumborClient.UrlBuilder
  alias ThumborClient.UrlBuilder

  ### Testing ThumborClient.URl
  test "#full_path with size and image" do
    assert UrlBuilder.full_path(%{width: 200, height: 300, image: "my-image.jpg"}) == "200x300/my-image.jpg"
  end

  test "#full_path with size, image and smart" do
    assert UrlBuilder.full_path(%{width: 200, height: 300, smart: true, image: "my-image.jpg"}) == "200x300/smart/my-image.jpg"
  end

  test "#trim without parameter" do
    assert UrlBuilder.trim([], %{}) == []
  end

  test "#trim with trim string" do
    assert UrlBuilder.trim([], %{trim: "bottom-right"}) == ["trim:bottom-right"]
  end

  test "#trim valid without value" do
    assert UrlBuilder.trim([], %{trim: true}) == ["trim"]
  end

  test "#meta parameter passing" do
    assert UrlBuilder.meta(["10x10"], %{meta: true}) == ["10x10", "meta"]
  end

  test "#meta without parameter" do
    assert UrlBuilder.meta(["10x10"], %{}) == ["10x10"]
  end

  test "#fit_in with empty value" do
    assert UrlBuilder.fit_in([], %{}) == []
  end

  test "#fit_in with invalid value" do
    assert UrlBuilder.fit_in([], %{fit: :invalid}) == []
  end

  test "#fit_in with fit_in" do
    assert UrlBuilder.fit_in([], %{fit: :fit_in}) == ["fit_in"]
  end

  test "#fit_in with adaptive_fit_in" do
    assert UrlBuilder.fit_in([], %{fit: :adaptive_fit_in}) == ["adaptive_fit_in"]
  end

  test "#fit_in with full_fit_in" do
    assert UrlBuilder.fit_in([], %{fit: :full_fit_in}) == ["full_fit_in"]
  end

  test "#fit_in with adaptive_full_fit_in" do
    assert UrlBuilder.fit_in([], %{fit: :adaptive_full_fit_in}) == ["adaptive_full_fit_in"]
  end

  test "#sizes param with width and height" do
    assert UrlBuilder.sizes([], %{width: 10, height: 20}) == ["10x20"]
  end

  test "#sizes width must be 0 if nil" do
    assert UrlBuilder.sizes([], %{height: 10}) == ["0x10"]
  end

  test "#sizes height must be 0 if nil" do
    assert UrlBuilder.sizes([], %{width: 10}) == ["10x0"]
  end

  test "#sizes width should be negative if flip" do
    assert UrlBuilder.sizes([], %{width: 10, flip: true, height: 10}) == ["-10x10"]
  end

  test "#sizes height should be negative if flop" do
    assert UrlBuilder.sizes([], %{width: 10, flop: true, height: 10}) == ["10x-10"]
  end

  test "#align with nil value" do
    assert UrlBuilder.align(["10x10"], %{}, :valign) == ["10x10"]
  end

  test "#align with center value" do
    assert UrlBuilder.align(["10x10"], %{valign: :center}, :valign) == ["10x10"]
  end

  test "#align with top value" do
    assert UrlBuilder.align(["10x10"], %{valign: :top}, :valign) == ["10x10", "top"]
  end

  test "#smart parameter passing" do
    assert UrlBuilder.smart(["10x10"], %{smart: true}) == ["10x10", "smart"]
  end

  test "#smart without parameter" do
    assert UrlBuilder.smart(["10x10"], %{}) == ["10x10"]
  end

  test "#filters with brightness" do
    assert UrlBuilder.filters([], %{filters: ["brightness(40)"]}) == ["filters:brightness(40)"]
  end

  test "#filters empty" do
    assert UrlBuilder.filters(["10x10"], %{filters: []}) == ["10x10"]
  end

  test "#filters without param" do
    assert UrlBuilder.filters(["10x10"], %{}) == ["10x10"]
  end

  test "#filters with multiple params" do
    assert UrlBuilder.filters(["10x10"], %{filters: ["blur(10)", "noise(20)", "rotate(45)"]}) ==
      ["10x10", "filters:blur(10):noise(20):rotate(45)"]
  end

  test "#image" do
    assert UrlBuilder.image(["10x10"], %{image: "sun.jpg"}) == ["10x10", "sun.jpg"]
  end

  test "#image raise error if option image is invalid" do
    assert_raise RuntimeError, ~r/image/, fn ->
      UrlBuilder.image([], %{})
    end
  end
end


mix.exs
defmodule ThumborClient.Mixfile do
  use Mix.Project

  def project do
    [
      app: :thumbor_client,
      version: "0.4.0",
      elixir: "~> 1.5",
      start_permanent: Mix.env == :prod,
      deps: deps(),
      package: package(),
      description: "Client to use Thumbor to crop images in Elixir"
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:ex_doc, "~> 0.11", only: :dev}
      # {:dep_from_hexpm, "~> 0.3.0"},
      # {:dep_from_git, git: "https://github.com/elixir-lang/my_dep.git", tag: "0.1.0"},
    ]
  end

  defp package do
    [
      maintainers: ["Tacnoman - Renato Cassino <renatocassino@gmail.com>"],
      licenses: ["MIT"],
      links: %{"GitHub" => "https://github.com/globocom/thumbor-client-ex"},
      files: [
        "lib/thumbor_client.ex",
        "lib/thumbor_client_url_builder.ex",
        "mix.exs",
        "README.md",
        "LICENSE",
        "CHANGELOG.md"
      ]
    ]
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure your application as:
#
#     config :thumbor_client, key: :value
#
# and access this configuration in your application as:
#
#     Application.get_env(:thumbor_client, :key)
#
# You can also configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/thumbor_client.ex
defmodule ThumborClient do
  @moduledoc """
  Documentation for ThumborClient.
  """

  alias ThumborClient.UrlBuilder

  def encrypt_to_thumbor(false, _path) do
    "unsafe"
  end

  @doc """
  Thumbor requires an specific encryption to work.
  The encryption is a HMAC-SHA1 using the secret key setted in thumbor server and the path.
  After encrypt with HMAC-SHA1, must be encrypted with Base64 url_encode.

  ## Examples

  iex> ThumborClient.encrypt_to_thumbor("123", "500/500/image.jpg")
  "VMKhXMULBn4h1UC52W3YliEgFfg="
  """
  def encrypt_to_thumbor(secret, path) do
    :crypto.hmac(:sha, secret, path)
    |> Base.url_encode64
  end

  @doc """
  Method to get the Thumbor Client
  This methods is a HOF to return the client of Thumbor

  ## Examples

  iex> client = ThumborClient.client("123")
  iex> client.(%{image: "image.jpg", width: 500, height: 400, smart: true})
  "rFZk5DrMK2hKAwVMJU4O4ZYDpeI=/500x400/smart/image.jpg"
  """
  def client(secret \\ false) do
    my_secret = secret

    fn(options) ->
      ThumborClient.generate(options, my_secret)
    end
  end

  @doc """
  Method to generate image passing parameters

  ## Examples
  iex> ThumborClient.generate(%{width: 200, height: 200, image: "my-image.jpg"}, "123")
  "gliOovhxLB8RGXinV2YT_W607lw=/200x200/my-image.jpg"

  OR

  iex> ThumborClient.generate(%{width: 200, height: 200, image: "my-image.jpg"})
  "unsafe/200x200/my-image.jpg"
  """
  def generate(options, secret \\ false) do
    full_path = UrlBuilder.full_path(options)
    encrypt_key = encrypt_to_thumbor(secret, full_path)

    "#{encrypt_key}/#{full_path}"
  end
end


lib/thumbor_client_url.ex
defmodule ThumborClient.Url do
  @moduledoc """
  Module to create url with params
  """

  def full_path(options) do
    []
    |> trim(options)
    |> meta(options)
    |> sizes(options)
    |> align(options, :halign)
    |> align(options, :valign)
    |> smart(options)
    |> image(options)
    |> Enum.join("/")
  end

  @doc """
  Removes surrounding space in images using top-left pixel color unless specified otherwise

  Unless specified trim assumes the top-left pixel color and no tolerance (more on tolerance below).

  To use it, just add a trim: true.

  If you need to specify the orientation from where to get the pixel color, just use trim: "top-left" for the top-left pixel color, for example

  Trim also supports color tolerance. The euclidian distance between the colors of the reference pixel and the surrounding pixels is used. If the distance is within the tolerance they'll get trimmed. For a RGB image the tolerance would be within the range 0-442. The tolerance can be specified like this: trim: "top-left:50"

  ## Examples

  iex> ThumborClient.Url.trim(["300x200"], %{trim: "bottom-left"})
  ["300x200", "trim:bottom-left"]

  iex> ThumborClient.Url.trim(["300x200"], %{trim: true})
  ["300x200", "trim"]
  """
  def trim(path, options) do
    case options[:trim] do
      nil -> path
      true -> path ++ ["trim"]
      trim -> path ++ ["trim:#{trim}"]
    end
  end

  @doc """
  Instead of get the image, get all meta data informations in image returning using json

  ## Examples

  iex> ThumborClient.Url.meta(["300x200"], %{meta: true})
  ["300x200", "meta"]
  """
  def meta(path, options) do
    case options[:meta] do
      nil -> path
      false -> path
      _ -> path ++ ["meta"]
    end
  end

  @doc """
  Add sizes to url image

  ## Examples

  iex> ThumborClient.Url.sizes([], %{width: 300, height: 200})
  ["300x200"]
  """
  def sizes(path, options) do
    width = options[:width]
    height = options[:height]

    path ++ ["#{width}x#{height}"]
  end

  @doc """
  Use orientation to crop image.

  Parameters:
  path: List with anothers attributes to build url
  options: Could be: :top, :left, :center, :right, :bottom
  orientation: Could be: :halign or :valign

  ## Examples

  iex> ThumborClient.Url.align(["300x200"], %{valign: :top}, :valign)
  ["300x200", "top"]
  """
  def align(path, options, orientation) do
    case options[orientation] do
      nil -> path
      false -> path
      :center -> path
      position -> path ++ [Atom.to_string(position)]
    end
  end

  @doc """
  Add url parameter smart to crop better images.
  Thumbor has algorithms to crop using facial recognition process.

  ## Examples

  iex> ThumborClient.Url.smart(["300x200"], %{smart: true})
  ["300x200", "smart"]
  """
  def smart(path, options) do
    case options[:smart] do
      nil -> path
      false -> path
      _ -> path ++ ["smart"]
    end
  end

  @doc """
  Add url image when receive options[:image]

  ## Examples

  iex> ThumborClient.Url.image([], %{image: "path/to/image.jpg"})
  ["path/to/image.jpg"]
  """
  def image(path, options) do
    path ++ [options[:image]]
  end
end


lib/thumbor_client_url_builder.ex
defmodule ThumborClient.UrlBuilder do
  @moduledoc """
  Module to build url with params
  """

  @doc """
  Get full path image passing options.
  """
  def full_path(options) do
    []
    |> trim(options)
    |> meta(options)
    |> fit_in(options)
    |> sizes(options)
    |> align(options, :halign)
    |> align(options, :valign)
    |> smart(options)
    |> filters(options)
    |> image(options)
    |> Enum.join("/")
  end

  @doc """
  Removes surrounding space in image using top-left pixel color unless specified otherwise

  Unless specified trim assumes the top-left pixel color and no tolerance (more on tolerance below).

  To use it, just add a trim: true.

  If you need to specify the orientation from where to get the pixel color, just use trim: "top-left" for the top-left pixel color, for example

  Trim also supports color tolerance. The euclidian distance between the colors of the reference pixel and the surrounding pixels is used. If the distance is within the tolerance they'll get trimmed. For a RGB image the tolerance would be within the range 0-442. The tolerance can be specified like this: trim: "top-left:50"

  ## Examples

  iex> ThumborClient.UrlBuilder.trim(["300x200"], %{trim: "bottom-left"})
  ["300x200", "trim:bottom-left"]

  iex> ThumborClient.UrlBuilder.trim(["300x200"], %{trim: true})
  ["300x200", "trim"]
  """
  def trim(path, options) do
    case options[:trim] do
      nil -> path
      true -> path ++ ["trim"]
      trim -> path ++ ["trim:#{trim}"]
    end
  end

  @doc """
  Instead of get the image, get all meta data informations in image returning using json

  ## Examples

  iex> ThumborClient.UrlBuilder.meta(["300x200"], %{meta: true})
  ["300x200", "meta"]
  """
  def meta(path, options) do
    case options[:meta] do
      nil -> path
      false -> path
      _ -> path ++ ["meta"]
    end
  end

  @doc """
  The fit argument specifies that the image should not be auto-cropped and auto-resized to be EXACTLY the specified size,
  and should be fit in an imaginary box of "E" width and "F" height, instead.

  Possible params: [:fit_in, :adaptive_fit_in, :full_fit_in, :adaptive_full_fit_in]
  Param in options: :fit

  ## Example

  iex> ThumborClient.UrlBuilder.fit_in(["200x200"], %{fit: :full_fit_in})
  ["200x200", "full_fit_in"]
  """
  def fit_in(path, options) do
    options_fit = [:fit_in, :adaptive_fit_in, :full_fit_in, :adaptive_full_fit_in]
    fit = Enum.find(options_fit, fn(key) ->
      options[:fit] == key
    end)

    if fit do path ++ [Atom.to_string(fit)] else path end
  end

  @doc """
  Add sizes to url image

  ## Examples

  iex> ThumborClient.UrlBuilder.sizes([], %{width: 300, height: 200})
  ["300x200"]
  """
  def sizes(path, options) do
    [width, height] = sizes_transform(options)
    path ++ ["#{width}x#{height}"]
  end

  def sizes_transform(options) do
    [
      size_transform(:width, options),
      size_transform(:height, options)
    ]
  end

  def size_transform(size_type, options) do
    fill_size_value(options[size_type])
    |> flip(options[if size_type == :width do :flip else :flop end])
  end

  def fill_size_value(size) do
    case size do
      nil -> 0
      _size -> size
    end
  end

  @doc """
  Function to flip values to int
  If second parameter is true: value * -1

  iex> ThumborClient.UrlBuilder.flip(10, true)
  -10

  iex> ThumborClient.UrlBuilder.flip(10, false)
  10
  """
  def flip(size, flip \\ false) do
    size * (if flip == true do -1 else 1 end)
  end

  @doc """
  Use orientation to crop image.

  Parameters:
  path: List with anothers attributes to build url
  options: Could be: :top, :left, :center, :right, :bottom
  orientation: Could be: :halign or :valign

  ## Examples

  iex> ThumborClient.UrlBuilder.align(["300x200"], %{valign: :top}, :valign)
  ["300x200", "top"]
  """
  def align(path, options, orientation) do
    case options[orientation] do
      nil -> path
      false -> path
      :center -> path
      position -> path ++ [Atom.to_string(position)]
    end
  end

  @doc """
  Add url parameter smart to crop better images.
  Thumbor has algorithms to crop using facial recognition process.

  ## Examples

  iex> ThumborClient.UrlBuilder.smart(["300x200"], %{smart: true})
  ["300x200", "smart"]
  """
  def smart(path, options) do
    case options[:smart] do
      nil -> path
      false -> path
      _ -> path ++ ["smart"]
    end
  end

  @doc """
  Adding filters to image. The option must be a List of strings.
  You can see all filters in https://github.com/thumbor/thumbor/wiki/Filters

  ## Examples

  iex> ThumborClient.UrlBuilder.filters(["300x300"], %{filters: ["rotate(30)", "brightness(40)"]})
  ["300x300", "filters:rotate(30):brightness(40)"]
  """
  def filters(path, options) do
    case options[:filters] do
      nil -> path
      [] -> path
      _filters -> path ++ [Enum.join(["filters"] ++ options[:filters], ":")]
    end
  end

  @doc """
  Add url image when receive options[:image]

  ## Examples

  iex> ThumborClient.UrlBuilder.image([], %{image: "path/to/image.jpg"})
  ["path/to/image.jpg"]
  """
  def image(path, options) do
    if options[:image] do
      path ++ [options[:image]]
    else
      raise "The option 'image' is required"
    end
  end
end
